### PR TITLE
AIML-1416 Add PMD and CPD static analysis with baseline gate

### PR DESCRIPTION
**Pull request by ChrisEdwards**
**Jira:** [AIML-1416](https://contrast.atlassian.net/browse/AIML-1416)

## Summary

Adds PMD and CPD static analysis to the Gradle check lifecycle, gated through per-module baseline files that track existing violations and prevent new ones. Four custom XPath rules ported from aiml-services enforce conventions already documented in CLAUDE.md. Changed-file PMD runs alongside changed-file coverage in both the pre-push hook and PR CI.

## Why This Change Exists

The aiml-services repo already has PMD catching issues like raw Mockito `any()` matchers, inline logging key literals, and missing `setCause()` in catch blocks. This repo had no equivalent, so those conventions were enforced only in review. PMD closes that gap. CPD catches copy-paste duplication that manual review misses.

The unified `staticAnalysisChangedBase` Gradle property was partially implemented but the CI workflow still passed the old `jacocoChangedBase` name. This PR finishes the unification so all changed-file analysis shares one property.

PR-posted coverage results (the third item in the bead) are deferred. The `common-github-actions/test-coverage` action lives in a private repo that public-repo workflows cannot reference. The existing `coverageSummary` to `GITHUB_STEP_SUMMARY` already provides coverage visibility on every CI run.

## The Big Picture

```mermaid
flowchart LR
    subgraph check["Gradle check lifecycle"]
        CS[Checkstyle] --> PMD[PMD + CPD]
        PMD --> BL[Baseline verify]
        BL --> PC[Positive controls]
        PC --> UT[Unit tests]
        UT --> JC[JaCoCo floors]
        JC --> CRAP[CRAP gate]
        CRAP --> PIT[PIT mutation]
    end
    subgraph changed["Changed-file gates (hook + CI)"]
        COV[Coverage verification]
        PMDCF[PMD verification]
    end
    check --> changed
```

Key decisions:

- **Per-module baselines** rather than a shared file. Two subprojects regenerating a shared baseline would clobber each other. Per-module files follow the existing `crap4j-baseline.json` precedent.
- **Dropped three aiml-services rules.** `UseJakartaNullable` (this repo uses Spring's `@Nullable`), `UseDecimalValidationForFloatingPoint` (no bean-validation usage), `CredentialFieldRequiresJsonIgnore` (no credential DTOs).
- **PMD 7.26.0** pinned in `gradle.properties` per ENTSEC-1742 soak policy (published 51 days ago).
- **buildSrc Java class** for baseline logic rather than inline Groovy. Keeps it testable and matches the `JacocoChangedFileCoverage` pattern.

### 1. PMD rulesets — **Focus**
The shared ruleset and main-only ruleset define what PMD checks. Four custom XPath rules target conventions from CLAUDE.md.
- `config/pmd/ruleset.xml` (new) — shared rules for main and test, including custom XPath rules
- `config/pmd/main-ruleset.xml` (new) — production-only `AvoidDuplicateLiterals`

### 2. Baseline verification logic — **Focus**
The core logic for reading PMD/CPD XML reports, comparing against baselines, and tightening.
- `buildSrc/src/main/java/.../PmdBaselineVerification.java` (new) — parse, verify, tighten
- `buildSrc/src/test/java/.../PmdBaselineVerificationTest.java` (new) — 16 tests covering PMD and CPD parsing, baseline read/write, new findings, stale entries, tighten

### 3. Gradle wiring — **Focus**
The largest single change. Adds PMD plugin, CPD task, 11 PMD-related tasks, positive controls, and check lifecycle wiring.
- `build.gradle` — PMD configuration, changed-file tasks, CPD, baseline verify/write/tighten, positive controls, check lifecycle

### 4. Hook and CI updates — **Skim**
Extends the existing changed-file gates to include PMD. Switches CI from `jacocoChangedBase` to `staticAnalysisChangedBase`.
- `scripts/git-hooks/pre-push` — adds `verifyPmdChangedFilesBaseline` alongside coverage
- `.github/workflows/build.yml` — adds PMD to the changed-file step, uploads PMD reports

<details>
<summary>Mechanical / low-risk changes (4 files)</summary>

- `gradle.properties` — adds `pmdVersion=7.26.0`
- `contrast-mcp-core/pmd-baseline.txt`, `contrast-mcp-stdio-app/pmd-baseline.txt` — generated initial baselines (52 and 38 entries)
- `docs/adr/0007-pmd-static-analysis-with-baselines.md` — ADR documenting the gate decision
- `CLAUDE.md`, `scripts/git-hooks/README.md` — documentation updates

</details>

## Testing

- `PmdBaselineVerificationTest` (16 tests) covers PMD report parsing, CPD report parsing, baseline reading (comments, blanks, missing files), new findings detection, stale entry detection, baseline write round-trip, and tighten operations.
- `make check` passes end to end with PMD, CPD, baseline verification, and positive controls.
- `make buildsrc-check` passes with the new buildSrc class meeting coverage floors.
- Positive controls verify that `FluentLogInCatchRequiresCause`, `AvoidInlineLoggingKeyLiteral`, and `AvoidRawMockitoAny` fire on synthetic source. `PreferLombokNoArgsConstructor` is not in the positive controls because the synthetic source would need specific class structure that conflicts with other test patterns.

## Risks / Rollout / Follow-ups

- **PR-posted coverage** deferred to a follow-up bead pending evaluation of public alternatives to the private `common-github-actions/test-coverage` action.
- Developers need to run `make install-hooks` to pick up the updated pre-push hook.
- Initial baselines are generated, not hand-curated. Some entries may represent quick wins for cleanup in future PRs.

---
Generated with Claude Code

[AIML-1416]: https://contrast.atlassian.net/browse/AIML-1416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: config/pmd/ruleset.xml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      language="java"
+      message="Use a typed Mockito matcher instead of raw any()."
+      class="net.sourceforge.pmd.lang.rule.xpath.XPathRule">
+    <description>Raw Mockito any() matchers hide type expectations in tests.</description>
+    <priority>3</priority>
+    <properties>
+      <property name="xpath">
+        <value>
+<![CDATA[
+//MethodCall[@MethodName = 'any' and count(ArgumentList/*) = 0]
```
Comment (ChrisEdwards) [inline config/pmd/ruleset.xml:52]: **[MEDIUM]** AvoidRawMockitoAny XPath matches any zero-arg `any()` method

The XPath `//MethodCall[@MethodName = 'any' and count(ArgumentList/*) = 0]` matches every zero-arg method named `any()`, not just Mockito's `ArgumentMatchers.any()`. If a future dependency or utility introduces an `any()` method, it will produce a false PMD finding. PMD XPath rules lack type resolution, so narrowing to Mockito-only calls requires checking the receiver expression chain or import context.

**Suggested fix:** Add a comment in the ruleset noting the intentional breadth, or narrow the XPath to check that the receiver chain includes `ArgumentMatchers`.

---
*Found by: Claude (correctness-adversarial-reviewer) + Codex (2/3 consensus)*

File: build.gradle [hunk 2]
```diff
@@ -463,6 +466,370 @@ subprojects {
...
+                    'cpd',
+                    '--minimum-tokens', '80',
+                    '--dir', cpdMainJavaDir.absolutePath,
+                    '--format', 'xml',
+                    '--report-file', cpdReportFile.get().asFile.absolutePath,
+                    '--no-fail-on-violation',
+                    '--skip-duplicate-files',
+                    '--ignore-literals',
+                )
+                ignoreExitValue = true
```
Comment (ChrisEdwards) [inline build.gradle]: **[MEDIUM]** CPD `ignoreExitValue` masks analyzer failures

The `cpdMain` task sets `ignoreExitValue = true`. CPD uses exit code 4 for "duplications found" (expected), but `ignoreExitValue` also swallows exit code 1 (internal error) and other failure codes. If CPD crashes or receives invalid config, no report is generated, and `verifyPmdBaseline` silently passes because it only checks reports that exist.

**Suggested fix:** Check `result.get().getExitValue()` and accept only 0 (clean) and 4 (duplications found), failing the task on any other exit code.

---
*Found by: Codex + Gemini (2/3 consensus)*

File: buildSrc/src/test/java/com/contrast/labs/build/PmdBaselineVerificationTest.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        Set.of("PMD|src/B.java|Rule2", "CPD|src/A.java,src/C.java", "PMD|src/A.java|Rule1");
+
+    PmdBaselineVerification.writeBaseline(baselineFile, findings, "my-module");
+
+    String content = Files.readString(baselineFile.toPath());
+    assertThat(content).startsWith("# PMD baseline for my-module.");
+    assertThat(content).contains("./gradlew :my-module:writePmdBaseline");
+
+    Set<String> roundTripped = PmdBaselineVerification.readBaseline(baselineFile);
+    assertThat(roundTripped).containsExactlyInAnyOrderElementsOf(findings);
```
Comment (ChrisEdwards) [inline buildSrc/src/test/java/com/contrast/labs/build/PmdBaselineVerificationTest.java:201]: **[MEDIUM]** writeBaseline test assertion ignores sort order

The test `writeBaseline_should_produce_sorted_entries_with_header` uses `containsExactlyInAnyOrderElementsOf(findings)`, which does not verify the sorted order the method name promises. The production code sorts via `TreeSet`, but if sorting were removed the test would still pass. Since baselines are committed to version control, sorted order matters for stable diffs.

**Suggested fix:** Replace with `containsExactly(...)` over a sorted expected list, or assert `isSortedAccordingTo(Comparator.naturalOrder())` on the non-comment lines read from the file.

---
*Found by: Claude (testing-reviewer)*

File: docs/adr/0007-pmd-static-analysis-with-baselines.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+The rules are ported from aiml-services, evaluated for applicability to this repo. Four custom XPath rules (AvoidRawMockitoAny, AvoidInlineLoggingKeyLiteral, FluentLogInCatchRequiresCause, PreferLombokNoArgsConstructor) are kept. Three rules specific to aiml-services conventions (UseJakartaNullable, UseDecimalValidationForFloatingPoint, CredentialFieldRequiresJsonIgnore) are dropped because this repo uses Spring's Nullable, has no bean-validation annotations, and has no credential DTOs with JSON serialization.
+
+## Considered options
+
+- **Fail on every finding from day one.** Rejected: the codebase has 90+ existing violations across two modules. Fixing them all upfront would mix cleanup with the gate itself, and fixing the wrong one could introduce regressions. The baseline records the debt and prevents new additions.
+- **Single shared baseline file.** Rejected: with two subprojects, a `writePmdBaseline` that rewrites one file from one module's reports would clobber the other module's entries. Per-module baselines (`contrast-mcp-core/pmd-baseline.txt`, `contrast-mcp-stdio-app/pmd-baseline.txt`) follow the per-module `crap4j-baseline.json` precedent.
```
Comment (ChrisEdwards) [inline docs/adr/0007-pmd-static-analysis-with-baselines.md:10]: **[MEDIUM]** ADR 0003 check lifecycle list does not include PMD/CPD

ADR 0003 enumerates what `check` runs (Spotless, Checkstyle, unit tests, JaCoCo, CRAP, PIT) but does not mention `verifyPmdBaseline` or `verifyPmdRuleControls`, which this PR wires into `check`. A reader relying on ADR 0003 will not know PMD/CPD are part of the gate.

Also affects: `scripts/git-hooks/README.md:68` — the "What counts as a pass" section explains only JaCoCo semantics but the hook now also runs PMD verification.

**Suggested fix:** Add PMD/CPD to the enumerated list in ADR 0003 and cross-reference ADR 0007. Add a paragraph to the git-hooks README explaining PMD's pass/fail rule.

---
*Found by: Docs staleness reviewer*

**Review by ChrisEdwards — COMMENTED** [review]
## Multi-model review

Well-structured PR that follows existing patterns (CRAP baselines, ArchUnit freeze store) and adds proper XML hardening. No CRITICAL or HIGH findings. Two MEDIUM findings have multi-engine consensus, both improvements rather than blockers. Two docs are slightly stale.

<details>
<summary>Minor notes (5)</summary>

- `Makefile:102` — Comment "what the pre-push hook does" is stale since hook now also runs PMD verification *(Claude correctness + standards)*
- `PmdBaselineVerificationTest.java:137` — Missing test for CPD clean report with zero duplications *(Claude testing)*
- `docs/adr/0003:15` — Deliberate exceptions list omits `verifyPmdChangedFilesBaseline` *(Docs staleness)*
- `PmdBaselineVerification.java:191` — DocumentBuilderFactory created per call; shared factory would avoid repeated feature negotiation at scale *(Claude security)*
- `build.gradle:205` — `pmdChangedBase` fallback property is introduced but never used by any caller *(Claude correctness)*
</details>

File: config/pmd/ruleset.xml (hunk 1 above)
Comment (ChrisEdwards) [reply config/pmd/ruleset.xml:52]: **Fixed** in 2b277b6 — added an XML comment above the XPath noting the intentional breadth and why narrowing isn't feasible (PMD XPath lacks type resolution).

File: build.gradle (hunk 2 above)
Comment (ChrisEdwards) [reply build.gradle]: **Fixed** in 2b277b6 — removed `ignoreExitValue = true` so Gradle's default behavior fails the task on any non-zero exit code. With `--no-fail-on-violation` already set, CPD exits 0 for both clean runs and runs with duplications. Only an internal error (exit 1) will now fail the task, preventing a silent pass.

File: buildSrc/src/test/java/com/contrast/labs/build/PmdBaselineVerificationTest.java (hunk 3 above)
Comment (ChrisEdwards) [reply buildSrc/src/test/java/com/contrast/labs/build/PmdBaselineVerificationTest.java:201]: **Fixed** in 2b277b6 — added `isSortedAccordingTo(Comparator.naturalOrder())` on the non-comment lines read from the file, verifying the sort order the method name promises. The existing round-trip assertion is kept alongside it.

File: docs/adr/0007-pmd-static-analysis-with-baselines.md (hunk 4 above)
Comment (ChrisEdwards) [reply docs/adr/0007-pmd-static-analysis-with-baselines.md:10]: **Fixed** in 2b277b6 — ADR 0003 now lists PMD/CPD (`verifyPmdBaseline`, `verifyPmdRuleControls`) in the `check` enumeration with a cross-reference to ADR 0007, and lists `verifyPmdChangedFilesBaseline` in the deliberate exceptions alongside JaCoCo. The git-hooks README was already updated in a prior commit.

**Comment by ChrisEdwards** [comment]
Addressing the minor notes from the multi-model review, all in 2b277b6:

- **#5 Makefile:102 stale hook comment** — **Fixed.** Comment now mentions PMD changed-file verification alongside coverage.
- **#6 Missing CPD clean report test** — **Fixed.** Added `cpdFindings_should_return_empty_for_clean_report`.
- **#7 ADR 0003 deliberate exceptions** — **Fixed.** `verifyPmdChangedFilesBaseline` added to the deliberate exceptions list in ADR 0003.
- **#8 DocumentBuilderFactory per call** — **Declined.** Runs at most ~4 times per build. Premature optimization for a build-time utility.
- **#9 pmdChangedBase fallback unused** — **Declined.** The property variable IS used (lines 521, 529). The `pmdChangedBase` Gradle property key is a forward-looking hook following the same `changedRefProperty` pattern established for JaCoCo.

**Review by Alex-Contrast — COMMENTED** [review]
Cross-model review of the PMD/CPD baseline gate, triaged against the repo at `2b277b6`. One finding, inline.

The gate mechanism itself checks out. `verifyPmdBaseline` throws on any non-baselined finding and `check` depends on it and on `verifyPmdRuleControls` (`build.gradle:830-832`), so it is the verifying task in the lifecycle rather than only the reporting one. CPD is gated on the same path, since `cpd.xml` lands in `reports/pmd` and `allFindings` dispatches it by filename. The property unification is complete, with no live consumer still reading `jacocoChangedBase`. The single finding is about retrieving the gate's output, not about the gate.

**Deliberately not findings**, recorded so they are not re-litigated:

- The XML parsing in `PmdBaselineVerification` is already hardened, `load-external-dtd` and both external-entity features off plus `FEATURE_SECURE_PROCESSING` on (`:191-196`). Checked because the input is a report generated from PR-authored source. Nothing to do.
- The PR body says 16 tests where the suite now has 18. Your own fix commit added `cpdFindings_should_return_empty_for_clean_report`, so the body is stale by one commit. Cosmetic, not worth a change.
- `newFindings_should_return_empty_when_all_findings_are_baselined` would pass under a comparison that returned "no new findings" unconditionally, but `newFindings_should_return_findings_not_in_baseline` would not, so the pair is sound. No change needed.

Nothing here duplicates your five-round self-review. All nine of those findings were checked against this one first, and none of them touches `.github/workflows/build.yml`.

**Open Questions:** none. Produced with model assistance and verified against the repository; no verdict is implied by this review.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -76,6 +77,7 @@ jobs:
           **/build/reports/jacoco/test/html/**
           **/build/reports/jacoco/test/jacocoTestReport.xml
           **/build/reports/pitest/**
+          **/build/reports/pmd/**
```
Comment (Alex-Contrast) [inline .github/workflows/build.yml:80]: **Low** (functional)

**Problem:** The `test-reports` artifact uploads `**/build/reports/pmd/**`, but the changed-file PMD tasks write to a *sibling* directory, `build/reports/pmd-changed/` (`build.gradle:475`, report locations set at `:543` and `:545`). The glob does not match a sibling, so `changed-main-files.html` and `changed-main-files.xml` are never uploaded. The positive-control reports at `build/reports/pmd-rule-controls/` (`:477`, written at `:788` and `:790`) are missed by the same glob.

**Impact:** `verifyPmdChangedFilesBaseline` runs in the PR path and `verifyPmdRuleControls` is wired into `check`, so either can fail a PR. When one does, the reports naming which rule fired on which line are absent from the artifact. This step is deliberately `if: ${{ !cancelled() }}` and its own comment says the reports are wanted "when the numbers and the HTML report are needed to diagnose the failure", so the gap defeats the step's stated purpose for two of the three PMD report sets. The gate itself still fails correctly, which is why this is Low.

**Fix:** Add both directories to the `path` list:

```yaml
          **/build/reports/pmd-changed/**
          **/build/reports/pmd-rule-controls/**
```

`**/build/reports/pmd*/**` would cover all three in one line, but the explicit list keeps the artifact contents readable and will not silently absorb a future `reports/pmd`-prefixed directory.

Checked while verifying the fix: `cpd.xml` lives *inside* `reports/pmd` (`build.gradle:577`), so it is already covered and neither addition displaces it. No other PMD or CPD task writes outside these three directories.

Comment (ChrisEdwards) [reply .github/workflows/build.yml:80]: Fixed in 6cc3ef7 on the tip branch (AIML-1416-pmd-rule-expansion). Added `**/build/reports/pmd-changed/**` and `**/build/reports/pmd-rule-controls/**` to the `test-reports` artifact path list in `build.yml`.

**Review by Alex-Contrast — APPROVED** [review]
